### ISSUE TITLE
Issues Swapping Rockbreaker with Shadow of the Destroyer

### ISSUE THREAD

**Issue by NaesKun** [0ceal0t/SkillSwap#8]
Might be related to this thread https://github.com/0ceal0t/SkillSwap/issues/5

Hey 0ceal0t! Naes again!

So I wanted to ask a quick question with a skill swap. I swapped Rockbreaker with Shadow of the Destroyer, animation goes off on all races except Lalafell, where the Lalafell just stands still and the VFX goes off. Do you know if there are different animation files for Lalafell for this specific instance? From my understanding, Lalafells just use Hyur .pap/.tmb that's just automatically rigged for their diminutive size. The fact nothing is happening is baffling me because I'm having no problems with any other skill swap but this one. Do you think you can highlight anything backend of what could be causing this?

**Comment by NaesKun** [comment]
Just a bump. This issue seems to be affecting any skill swaps with EW skills. EW skills don't carry over animations properly to Lalafell races it seems. Another mod that puts Scattershot over Spread Shot and Autocrossbow is running into this issue as well.

**Comment by NaesKun** [comment]
Further testing revealed its not an issue with EW skills, its just an issue with the skill itself. Rockbreaker swapped with any Monk AoE animation will be broken on Lalafell. Don't think there's a way around this. Closing ticket.

**Comment by 0ceal0t** [comment]
that's really strange. Sorry for not giving this a proper look yet, since SkillSwap has a lot of weird edge cases. I'll see if I can dig around in the pap files and find out why this is only an issue for lalafell

**Comment by NaesKun** [comment]
It affect skills randomly. Here's a list of positives I've found:

- Rockbreaker swapped with any animation causes Lalafell to just stand there while the VFX plays.
- Flint Strike swapped with any animation causes Lalafell to just stand there while the VFX plays.
- Rampart swapped with any skill causes both Lalafell and Roegadyn/Hrothgar males to just stand there while the VFX plays.
- Hypercharge swapped with any skill casues Lalafell to just stand there while the VFX plays.

Most of the time it looks like specific skills don't work for Lalafells after SkillSwap is applied. My guess is that it has something to with the Skill's logic and if the skill doesn't have the original .pap/.tmb file to reference, then it just doesn't work on Lalafell.

Would be nice to get a confirmation on the cause of this issue so I can get a bit of sanity back.

**Comment by NaesKun** [comment]
Heya 0ceal0t. Just wanted to know if there's any update on this issue. Has a cause been identified or can we just chalk this up to an unsolvable issue?

**Comment by 0ceal0t** [comment]
no update yet, sorry. I've been really busy with other stuff so I haven't had a chance to look to far into it. For now you can consider it unsolvable

**Comment by 0ceal0t** [comment]
ok I tested it out just for my own peace of mind. SkillSwap has some weird edge cases because it tries to do pap animation id renaming automatically to prevent conflicts. Sometimes it doesn't work, though.

However, it seems to work fine when swapped in VFXEditor, so I'll recommend doing that instead. Just know that you might have to change some of the pap animation ids in the `.pap` and `.tmb` files manually (this is just for illustrative purposes, for Rockbreaker > Arm of the Destroyer you can just swap the `.tmb` without any extra work)

![image](https://user-images.githubusercontent.com/18051158/152611068-81062440-9567-4236-ab6e-4ae6e5dd4af4.png)

**Comment by NaesKun** [comment]
Gotcha. So the .pap files will need to be manually adjusted because of the way the IDs are used by the game.

Man this is gonna be so much work, but thanks for clearing this up. This should help future modders too with their animation issues.

**Comment by NaesKun** [comment]
Sorry to reopen, I'm trying to understand the logic myself and I'm not able to grasp it.

I tried following the instructions word for word in the above image when I tried changing Hypercharge to Rapid Fire's animation and it still breaks down on Lalafell where the Lalafell just stands there without the animation. I'm not sure if my logic is incorrect or if there's a hard technical limitation for what can be animation swapped through VFX editor.

Can you ELI5 and explain step by step how I would make the animation work for Lalafell so I make sure I have the correct logic moving forward?

**Comment by 0ceal0t** [comment]
ok, I found the solution. You have to set it up like this: 

![image](https://user-images.githubusercontent.com/18051158/152694039-38c622d2-62da-4851-bb22-e351a9dccbdd.png)

**Comment by 0ceal0t** [comment]
the path is `chara/human/c1101/animation/a0001/bt_common/ability/2gn_machin/abl012.pap`, which is the `.pap` for hypercharge specifically for Lalafel.

You can find this using Explorer, since male Lalefel have the skeleton id `c1101` and hypercharge is `abl012`

![image](https://user-images.githubusercontent.com/18051158/152694084-69cde1f1-e08b-4bef-a553-216aef9745e3.png)


I'll work on trying to get this integrated into VFXEditor, or at least writing up some documentation on it

**Comment by NaesKun** [comment]
Thanks a lot 0ceal0t. This will help with so many fringe cases in case SkillSwap and your first fix don't go through.

Also really big thanks on the second image, I can actually use Explorer/Godbert and know what to look for.

**Comment by DarkPoseidon** [comment]
Thanks a lot Oceal0t for looking further into this Lala problem. Do you know what is the Female Lala skeleton id? I can't seem to find that out.

I am trying to replace Inner Chaos with Plentiful Harvest through the VFX editor like this for a Female Lala.

![image](https://user-images.githubusercontent.com/96473079/152746284-58ab6ff9-c82f-4628-88ae-a763231019ff.png)

**Comment by NaesKun** [comment]
Just a quick question, 

> 

AFAIR, Female Lala's use the same animation files as male Lala's. 

Also, I would strongly recommend against using the Lala files listed in the VFX Editor. For me I got weird edge cases where Roe and Hyur animations stop working.

I recommend following 0ceal0t's guide exactly:

1. For Loaded TMB and TMB Being Replaced, use the generic end file or the Hyur end file for both entries.
2. For Loaded PAP use the generic file or the Hyur .pap file.
3. For PAP being replaced, put in the exact file path for the Lalafell version of the animation. You will have to use FFXIV Explorer to find the direct path to the file.

Specifically, I use TexTools. If you use Pneumbra ignore these instructions entirely. Otherwise you need to do additional steps to make sure you get this working.
1. Export the file out to TexTools from VFX Editor. Make sure you select all the option available.
2. Go into SkillSwap and swap the targeted skill with the desired skill as normal and save the mod.
3. Exit the game, open the Lalafell mod you exported from VFX editor and import both the .pap and the .tmb.
4. Open the SkillSwap mod you created and ONLY IMPORT THE PAP FILE. Do not import the tmb or you will break animations.
5. Open the game again and test, should work for all races now.

Its a long tricky process at first, but once you get the hang of it, its super easy.

**Comment by NaesKun** [comment]
@0ceal0t Quick question. Do you know what the Roe/Hroth Skeleton ID is? I'm trying to fix some other mods as well and the problem extends to Roe skeletons as well.

**Comment by 0ceal0t** [comment]
I'm curious why using the paths generated by vfxeditor breaks other skeletons, though. that's really weird

**Comment by NaesKun** [comment]
@0ceal0t Could also just be how TexTools handles the files when we import. From my experience, it looks like TexTools does not like it when multiple of the same .tmbs are applied, especially if they're tied to animations.

I wouldn't worry too much about it if you're busy. The fix works great currently and I'll expand this to Roe/Lala when I explore adjustments to Rampart and let you know if there are any issues there, but I don't think there should be if we follow this logic correctly.

**Comment by NaesKun** [comment]
@0ceal0t Did some testing and isolated the issue.

I followed my logic in my reply above (https://github.com/0ceal0t/SkillSwap/issues/8#issuecomment-1031565104) where I was able to successfully create working animations for all races for the following skill swaps:

Rockbreaker to Shadow of the Destroyer
Flint Strike to Rockbreaker

Individually, these files work perfectly on their own. However, as soon as I apply them together through TexTools, shit starts breaking down. All the animations break on Flint Strike except for Lalafell which works properly. My guess is that it has to do with the way SkillSwap handled the animation swaps.

My next test would be to adjust all the animations manually for both skills and see if this resolves the issue.

**Comment by NaesKun** [comment]
Hmm, actually it could be an ID issue where both animations are tied to Rockbreaker and pulling off some wonky shit. Might be worth it to try creating unique .pap IDs for the Flint Strike changes to see if that resolves the issue.

**Comment by NaesKun** [comment]
Tried it just now and it breaks even before export (animations stop working).

![image](https://user-images.githubusercontent.com/93843234/152826599-6d901c4e-4f5f-4400-adc1-f5486c5e3671.png)

Not sure if I'm following the right logic, I tried combining your two steps above and it ends up just not working.

**Comment by NaesKun** [comment]
Going to need to think about alternative approaches, I think the common culprit is SkillSwap.

**Comment by 0ceal0t** [comment]
![image](https://user-images.githubusercontent.com/18051158/152829333-35494c51-1909-4fae-8481-b17a1ba6aae0.png)

I didn't have any issues doing this:

**Comment by NaesKun** [comment]
@0ceal0t I found a solution, which was manually changing the animation for both Hyur and Lalafell through VFX editor. There weren't any naming conflicts like you mentioned before so it looks like the bugs are tied to the way SkillSwap handles the whole animation swapping.

I'm going to have to test further with all my other mods to see if there's any other mods that might conflict with this, but for now it looks okay. I'll provide further results of testing in this threat. 

@DarkPoseidon I revised what I did. Its working so far, but this is still in testing so YMMV:

1. For Loaded TMB and TMB Being Replaced, use the "end .tmb" file type for what you want to do.
2. For Loaded PAP use the Hyur end TMB file or Lalafell end TMB file.
3. For PAP being replaced, put in the exact file path for the Lalafell OR Hyur version of the animation and make sure it matches what you originally put into the Loaded PAP entry. You will have to use FFXIV Explorer to find the direct path to the file (should be under the 040000.index file).
4. Export the file out to TexTools from VFX Editor. Make sure you select all the options available.
5. Repeat above steps for both Lalafell and Hyur.
6. Exit the game, open the Lalafell mod you exported from VFX editor and import both the .pap and the .tmb.
7. Open the SkillSwap mod you created and ONLY IMPORT THE PAP FILE. Do not import the tmb or you will break animations.
8. Open the game again and test, should work for all races now.

**Comment by 0ceal0t** [comment]
I'm not sure if that whole complicated process is needed @NaesKun 
![flint](https://user-images.githubusercontent.com/18051158/152882720-5390ed92-2f74-4e80-973f-a75611f0fd83.PNG)

**Comment by 0ceal0t** [comment]
[rockbreaker_flint_strike.zip](https://github.com/0ceal0t/SkillSwap/files/8019169/rockbreaker_flint_strike.zip)

here's the mod which came from that if you want to try it yourself

**Comment by DarkPoseidon** [comment]
Sorry to come back to this Lala issue again. I tried the simplified process of creating a new .pap for each racial but I still can't seem to make the animation work.

I am trying to replace Storm's Eye (warrior) into Void Reaping (reaper). And Inner Chaos (warrior) into Plentiful Harvest (reaper).

That's the exported ttmp file that I imported to Penumbra and the way I set up the vfx edit.
@0ceal0t @NaesKun Could I trouble either of you to send me the ttmp files that is correct?

[Storms Eye and Inner Chaos.zip](https://github.com/0ceal0t/SkillSwap/files/8047051/Storms.Eye.and.Inner.Chaos.zip)

![Picture1](https://user-images.githubusercontent.com/96473079/153558346-dff058e6-120d-43ad-836c-37090462628e.png)

**Comment by NaesKun** [comment]
> Sorry to come back to this Lala issue again. I tried the simplified process of creating a new .pap for each racial but I still can't seem to make the animation work.
> 
> I am trying to replace Storm's Eye (warrior) into Void Reaping (reaper). And Inner Chaos (warrior) into Plentiful Harvest (reaper).
> 
> That's the exported ttmp file that I imported to Penumbra and the way I set up the vfx edit. @0ceal0t @NaesKun Could I trouble either of you to send me the ttmp files that is correct?
> 
> [Storms Eye and Inner Chaos.zip](https://github.com/0ceal0t/SkillSwap/files/8047051/Storms.Eye.and.Inner.Chaos.zip)
> 
> ![Picture1](https://user-images.githubusercontent.com/96473079/153558346-dff058e6-120d-43ad-836c-37090462628e.png)

I don't use the built in target path in the editor, the results have been wildly inconsistent in my experience.

I do a manual path and find the animation path directly in XIV explorer and punch that in instead. 0ceal0t might have a solution for this issue though that I might be ignorant about.
